resources/assets/lib/news-announcements/main.tsx

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import Img2x from 'components/img2x';
+import NewsAnnouncementJson from 'interfaces/news-announcement-json';
+import { action, autorun, IReactionDisposer, makeObservable, observable } from 'mobx';
+import { observer } from 'mobx-react';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+
+const bn = 'news-announcements';
+const itemBn = 'news-announcement';
+
+interface Props {
+  announcements: NewsAnnouncementJson[];
+}
+
+@observer
+export default class Main extends React.Component<Props> {
+  @observable private announcementWidth = 0;
+  @observable private readonly containerRef = React.createRef<HTMLDivElement>();
+  @observable private index = 0;
+  private registerResizeObserverDisposer?: IReactionDisposer;
+  private resizeObserver?: ResizeObserver;
+  private rotateAnnouncementTimer?: NodeJS.Timer;
+  @observable private readonly topRef = React.createRef<HTMLDivElement>();
+
+  private get length() {
+    return this.props.announcements.length;
+  }
+
+  constructor(props: Props) {
+    super(props);
+
+    makeObservable(this);
+  }
+
+  componentDidMount() {
+    this.setRotateAnnouncementTimer();
+
+    this.registerResizeObserverDisposer = autorun(() => {
```

Comment:
use `disposeOnUnmount`; actually, this shouldn't even be `autorun` because the refs shouldn't be observable - they will either be there or not be there after render.